homeassistant/components/monzo/__init__.py

```diff
@@ -27,18 +63,157 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+                except InvalidMonzoAPIResponseError:
+                    _LOGGER.error("Error during webhook registration")
+                else:
+                    self.entry.async_on_unload(self.unregister_webhooks)
+
+    async def unregister_old_webhooks(self, coordinator: MonzoCoordinator) -> None:
+        """Unregister any old webhooks associated with this host."""
+        host = (
+            CLOUDHOOK_HOST
+            if cloud.async_active_subscription(self.hass)
+            else get_url(self.hass)
+        )
+        await coordinator.api.user_account.unregister_webhooks(host)
+
+    async def _async_cloudhook_generate_url(self, webhook_id: str) -> str:
+        """Generate the full URL for a webhook_id."""
+        webhook_url = await cloud.async_create_cloudhook(self.hass, webhook_id)
+        self.entry.runtime_data.cloudhook_urls.add(webhook_url)
+        return webhook_url
+
+    async def unregister_webhooks(self) -> None:
+        """Unregister all Monzo webooks."""
+        coordinator: MonzoCoordinator = self.entry.runtime_data.coordinator
+
+        async_dispatcher_send(
+            self.hass,
+            f"signal-{DOMAIN}-webhook-None",
+            {"type": "None", "data": {WEBHOOK_PUSH_TYPE: WEBHOOK_DEACTIVATION}},
+        )
+        while self.entry.runtime_data.webhook_ids:
+            webhook_id = self.entry.runtime_data.webhook_ids.pop()
+            _LOGGER.debug("Unregister Monzo webhook (%s)", webhook_id)
+            webhook_unregister(self.hass, webhook_id)
+
+            if cloud.async_active_subscription(self.hass):
+                try:
+                    _LOGGER.debug("Removing Monzo cloudhook (%s)", webhook_id)
+                    await cloud.async_delete_cloudhook(self.hass, webhook_id)
+                except cloud.CloudNotAvailable:
+                    pass
+        await self.unregister_old_webhooks(coordinator)
+
+
+async def async_handle_webhook(
+    hass: HomeAssistant, webhook_id: str, request: Request
+) -> None:
+    """Handle webhook callback."""
+    try:
+        data = await request.json()
+    except ValueError as err:
+        _LOGGER.error("Error in data: %s", err)
+        return None
+
+    _LOGGER.debug("Got webhook data: %s", data)
+
+    event_type = data.get("type")
+
+    if event_type == EVENT_TRANSACTION_CREATED:
+        async_send_event(hass, event_type, data.get("data"))
+    else:
+        _LOGGER.debug("Got unexpected event type from webhook: %s", event_type)
+
+
+def async_send_event(hass: HomeAssistant, event_type: str, data: dict) -> None:
+    """Send events."""
+    _LOGGER.debug("%s: %s", event_type, data)
+    async_dispatcher_send(
+        hass,
+        f"signal-{DOMAIN}-webhook-{event_type}",
+        {"type": event_type, "data": data},
+    )
+
+    event_data = {"type": event_type, "data": data, "account_id": data["account_id"]}
+
+    hass.bus.async_fire(
+        event_type=MONZO_EVENT,
+        event_data=event_data,
+    )
```

Comment:
Let's not fire on the event bus

Reply:
I'm not 100% sure how to do this without the event. Is it a PluggableAction I want instead?

Reply:
You can use the dispatcher